readme.md
-------

# ALIAS BOARDS v2.9

[**Written by SleePy**](http://www.simplemachines.org/community/index.php?action=profile;u=29084) - [**Taken Over By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913)

-------

## Install Requirements
This mod was tested on **SMF 2.0.11**, but should work on **SMF 2.1 Beta 3**, as well as **SMF 2.0 and up**.  **SMF 2.1 Beta 1, 2.1 Beta 2**, and **SMF 1.x** will not and will not be supported.

If you are using [Split Forum mod](http://custom.simplemachines.org/mods/index.php?mod=3730), this mod should be installed **BEFORE** the Split Forum mod!

## Introduction
This Mod allows you to Alias/Symlink or more commonly known for windows users as shortcuts.
This emulates another board by copying its data into another location. This will show on the MessageIndex as well.

## Admin Settings
Under **Admin** => **Forum** => **Boards**, functionality has been added to the Add and Edit Board UIs in order to add the ability to create, remove and update the aliases to boards 

## Translators
o Spanish Latin: [Rock Lee [BC]](http://www.simplemachines.org/community/index.php?action=profile;u=322597)

## Additional Information
-  [This mod on SMF Customization Site](http://custom.simplemachines.org/mods/index.php?mod=1024)
-  [This mod at XPtsp.com](

## License
Copyright (c) 2017, Douglas Orend, Kirk Sykora
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changelog.txt
v2.9 - February 25th, 2019
o Updated for SMF 2.1 RC 1.
o Removed support for SMF 2.1 Beta 3.

v2.8 - October 9th, 2018
o No functionality change.
o Updated documentation to point to new website and GitHub repository.

v2.7 - February 10th, 2018
o No functionality changes.
o Added missing copyright owner to [b]LICENSE[/b] file.
o Added translators to mod documentation file [b]readme.txt[/b].

v2.6 - July 5th, 2017
o Updated for SMF 2.1 Beta 3.
o Removed support for SMF 2.1 Beta 1 and 2.

v2.5 - January 28th, 2017
o Fixed error that occurs when creating a new post.
o Add a LICENSE file for the current license, 2-clause BSD.

v2.4 - April 19th, 2016
o Fixed several undefined element errors when used in conjunction with Split Forum mod.

v2.3 - April 15th, 2016
o Added code to make aliases within same category show up properly.
o Added Spanish Latin translation by [url=http://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee [BC][/url].

v2.2 - January 17th, 2016
o Added code to automatically add the proper board redirect to the boards database.
o Added code to remove the board information from aliased boards when unaliasing a board.
o Added code to remove board redirect when unchecking alias board option in Edit Board UI.
o Fixed issue where an alias can't be created with 1 board already in existance.
o Modified [b]db_install.php[/b] to create board redirects for all existing aliases.
o Removed [b]db_uninstall.php[/b] from mod, which is responsible for removing board aliases.
o SMF 2.1: Fixed undefined index error in Edit Board UI.

v2.1 - January 2nd, 2016
o Removed option to create alias if one or fewer boards are present.

v2.0 - December 4th, 2015
o Complete rewrite to enable proper sorting and placement of aliased boards.
o This is NOT compatible with version 1.x of this mod!

v1.4 - April 6th, 2015
o Updated for SMF 2.1 Beta 1

v1.3 - January 16th, 2015
o License change to Simplified BSD.  No functionality change.

v1.2 - October 31st, 2014
+ Fixed the Manage Board UI using the changes Bugo posted

Version 1.1.8
! Updated for SMF 2.0 Final and added BSD License

Version 1.1.7
! Updated for SMF 2.0 RC4

Version 1.1.6
! Updated for SMF 2.0 RC3

Version 1.1.5
! Database install/removal was broken in SMF 2.0 RC2

Version 1.1.4
! Updated to support SMF 2.0 RC1

Version 1.1.3
+ Took use of SMF Caching (SMF 2.0 + only) and adding a time setting for it.
! Updated Support for of only SMF 2.0 Beta 4
! Fix issues with Custom Board sort

Version 1.1.2
! Added Support for SMF 2.0 Beta 3 Public
! Dropped support for old versions of 2.0 Betas
+ Rewrote database installer/remover to work with other databases.

Version 1.1.1
! Fixed to work with 2.0 Beta 2.1

Version 1.1
! Fixed to work with 2.0 Beta 2

Version 1.0
! Release

install.php
<?php
error_reporting(E_ALL);

// Hopefully we have the goodies.
if (file_exists(dirname(__FILE__) . '/SSI.php') && !defined('SMF'))
{
	$using_ssi = true;
	require_once(dirname(__FILE__) . '/SSI.php');
}
elseif (!defined('SMF'))
	exit('<b>Error:</b> Cannot install - please verify you put this in the same place as SMF\'s index.php.');

global $db_prefix, $modSettings, $func, $smcFunc;
$modSettings['disableQueryCheck'] = true;

// SleePy is lazy, so I will just check to see if db_query exists
if(!function_exists('db_query'))
{
	db_extend('Packages');
	function db_query($query, $file, $line)
	{
		global $smcFunc;
		return $smcFunc['db_query']('', $query, array('db_error_skip' => true));
	}
}
// All work here is for back support for SMF 1.1, It is easier to support 2.0 and backport.
else
{
	$smcFunc = $func;
	$smcFunc['db_num_rows'] = 'mysql_num_rows';
	$smcFunc['db_free_result'] = 'mysql_free_result';
	$smcFunc['db_fetch_assoc'] = 'mysql_fetch_assoc';
	$smcFunc['db_list_columns'] = 'mysql_show_columns';
	$smcFunc['db_add_column'] = 'mysql_create_columns';

	// Quickly emulate these functions.
	function mysql_show_columns($table_name)
	{
		global $smcFunc, $db_prefix;

		$result = db_query("SHOW FIELDS FROM {$table_name}", __FILE__, __LINE__);;
		$columns = array();
		while ($row = $smcFunc['db_fetch_assoc']($result))
			$columns[] = $row['Field'];
		return $columns;
	}
	function mysql_create_columns($table_name, $column_info)
	{
		global $db_prefix;

		return db_query('ALTER TABLE ' . $table_name . '
			ADD ' . $column_info['name'] . ' ' . $column_info['type'] . ' ' . (empty($column_info['null']) ? 'NOT NULL' : '') . ' ' .
		(empty($column_info['default']) ? '' : 'default \'' . $column_info['default'] . '\'') . ' ' .
		(empty($column_info['auto']) ? '' : 'auto_increment') . ' ', __FILE__, __LINE__);
	}
}

// Fields to add
$new_fields = array(
	'alias_cat' => array('name'=> 'alias_cat', 'type'=>'VARCHAR(255)'),
	'alias_child' => array('name'=> 'alias_child', 'type'=>'VARCHAR(255)'),
);

// Do you already got my goodies!
$table_columns = $smcFunc['db_list_columns']($db_prefix . "boards");

// Do the loopy, loop, loe.
foreach ($new_fields as $column_name => $column_attributes)
{
	// Dang, I guess I must share.
	if(!in_array($column_name, $table_columns))
		$smcFunc['db_add_column']($db_prefix . "boards", $column_attributes);
}

if(!empty($using_ssi))
	echo 'If no errors, Success!';
?>

db_install.php
<?php
global $db_prefix, $modSettings, $func, $smcFunc;

// Hopefully we have the goodies.
if (file_exists(dirname(__FILE__) . '/SSI.php') && !defined('SMF'))
{
	$using_ssi = true;
	require_once(dirname(__FILE__) . '/SSI.php');
}
elseif (!defined('SMF'))
	exit('<b>Error:</b> Cannot install - please verify you put this in the same place as SMF\'s index.php.');
db_extend('Packages');

// Add the "alias_board" column to the boards table
$smcFunc['db_add_column'](
	'{db_prefix}boards', 
	array(
		'name' => 'alias_board', 
		'size' => 5, 
		'type' => 'smallint', 
		'null' => false, 
		'default' => 0
	)
);

// Change the redirects so that when the mod is uninstalled, the board "alias" becomes a redirect:
$smcFunc['db_query']('', '
	UPDATE {db_prefix}boards
	SET redirect = CONCAT("' . $boardurl . '/index.php?board=", alias_board, ".0") 
	WHERE alias_board > {int:no_alias}',
	array(
		'no_alias' => 0,
	)
);

if(!empty($using_ssi))
	echo 'If no errors, Success!';
?>

remove.php
<?php
error_reporting(E_ALL);

// Hopefully we have the goodies.
if (file_exists(dirname(__FILE__) . '/SSI.php') && !defined('SMF'))
{
	$using_ssi = true;
	require_once(dirname(__FILE__) . '/SSI.php');
}
elseif (!defined('SMF'))
	exit('<b>Error:</b> Cannot install - please verify you put this in the same place as SMF\'s index.php.');

global $db_prefix, $modSettings, $func, $smcFunc;
$modSettings['disableQueryCheck'] = true;

// SleePy is lazy, so I will just check to see if db_query exists
if(!function_exists('db_query'))
{
	db_extend('Packages');
	function db_query($query, $file, $line)
	{
		global $smcFunc;
		return $smcFunc['db_query']('', $query, array('db_error_skip' => true));
	}
}
// All work here is for back support for SMF 1.1, It is easier to support 2.0 and backport.
else
{
	$smcFunc = $func;
	$smcFunc['db_num_rows'] = 'mysql_num_rows';
	$smcFunc['db_free_result'] = 'mysql_free_result';
	$smcFunc['db_fetch_assoc'] = 'mysql_fetch_assoc';
	$smcFunc['db_list_columns'] = 'mysql_show_columns';
	$smcFunc['db_add_column'] = 'mysql_create_columns';

	// Quickly emulate these functions.
	function mysql_show_columns($table_name)
	{
		global $smcFunc, $db_prefix;

		$result = db_query("SHOW FIELDS FROM {$table_name}", __FILE__, __LINE__);;
		$columns = array();
		while ($row = $smcFunc['db_fetch_assoc']($result))
			$columns[] = $row['Field'];
		return $columns;
	}
	function mysql_create_columns($table_name, $column_info)
	{
		global $db_prefix;

		return db_query('ALTER TABLE ' . $table_name . '
			ADD ' . $column_info['name'] . ' ' . $column_info['type'] . ' ' . (empty($column_info['null']) ? 'NOT NULL' : '') . ' ' .
		(empty($column_info['default']) ? '' : 'default \'' . $column_info['default'] . '\'') . ' ' .
		(empty($column_info['auto']) ? '' : 'auto_increment') . ' ', __FILE__, __LINE__);
	}
}

// Fields to add
$new_fields = array(
	'alias_cat' => array('name'=> 'alias_cat', 'type'=>'VARCHAR', 'size' => '255'),
	'alias_child' => array('name'=> 'alias_child', 'type'=>'VARCHAR', 'size' => '255'),
);

// Do you already got my goodies!
$table_columns = $smcFunc['db_list_columns']($db_prefix . "boards");

// Do the loopy, loop, loe.
foreach ($new_fields as $column_name => $column_attributes)
{
	// Dang, I guess I must share.
	if(!in_array($column_name, $table_columns))
		$smcFunc['db_remove_column']($db_prefix . "boards", $column_attributes['name'], array('no_prefix' => TRUE));
}

if(!empty($using_ssi))
	echo 'If no errors, Success!';
?>